src/Microsoft.AspNetCore.Mvc.Testing/WebHostBuilderExtensions.cs

```diff
@@ -0,0 +1,112 @@
+        /// <param name="args">The arguments passed to CreateDefaultBuilder, typically from <c>Program.Main(string [] args)</c>.</param>
+        /// <returns>An <see cref="IWebHostBuilder"/> instance.</returns>
+        public static IWebHostBuilder FromStartup<TStartup>(string[] args)
+        {
+            var factory = typeof(TStartup).Assembly.EntryPoint.DeclaringType.GetMethod(CreateDefaultBuilder, BindingFlags.Static | BindingFlags.Public);
+            if (factory == null)
+            {
+                return null;
+            }
+
+            return (IWebHostBuilder)factory.Invoke(null, new object[] { args });
+        }
+
+        /// <summary>
+        /// Configures <see cref="ApplicationPartManager"/> to include the default set
+        /// of <see cref="ApplicationPart"/> provided by <see cref="DefaultAssemblyPartDiscoveryProvider"/>.
+        /// </summary>
+        /// <param name="builder">The <see cref="IWebHostBuilder"/> to configure.</param>
+        /// <typeparam name="TStartup">The Startup type.</typeparam>
+        /// <returns>An instance of this <see cref="IWebHostBuilder"/></returns>
+        public static IWebHostBuilder UseApplicationAssemblies<TStartup>(this IWebHostBuilder builder)
+        {
+            var depsFileName = $"{typeof(TStartup).Assembly.GetName().Name}.deps.json";
+            var depsFile = new FileInfo(Path.Combine(AppContext.BaseDirectory, depsFileName));
+            if (!depsFile.Exists)
+            {
+                throw new InvalidOperationException($"Can't find'{depsFile.FullName}'. This file is required for functional tests " +
+                    "to run properly. There should be a copy of the file on your source project bin folder. If thats not the " +
+                    "case, make sure that the property PreserveCompilationContext is set to true on your project file. E.g" +
+                    "'<PreserveCompilationContext>true</PreserveCompilationContext>'." +
+                    $"For functional tests to work they need to either run from the build output folder or the {Path.GetFileName(depsFile.FullName)} " +
+                    $"file from your application's output directory must be copied" +
+                    "to the folder where the tests are running on. A common cause for this error is having shadow copying enabled when the" +
+                    "tests run.");
+            }
+
+            var assemblies = DefaultAssemblyPartDiscoveryProvider
+                .DiscoverAssemblyParts(typeof(TStartup).Assembly.GetName().Name)
+                .Select(s => ((AssemblyPart)s).Assembly)
+                .ToList();
+
+            builder.ConfigureServices(services =>
+                services.AddSingleton<IStartupConfigureServicesFilter>(new ApplicationAssembliesStartupServicesFilter(assemblies)));
+
+            return builder;
+        }
+
+        /// <summary>
+        /// Configures the application content root.
```

Comment:
We should probably document how this works i.e. we look for the first directory containing the pattern `solutionName` starting from the directory containing the application.